component/main.jsonnet

```diff
@@ -1,10 +1,110 @@
 // main template for csi-driver-smb
+local com = import 'lib/commodore.libjsonnet';
+local espejo = import 'lib/espejo.libsonnet';
 local kap = import 'lib/kapitan.libjsonnet';
 local kube = import 'lib/kube.libjsonnet';
 local inv = kap.inventory();
 // The hiera parameters for the component
 local params = inv.parameters.csi_driver_smb;
 
-// Define outputs below
+local pvs = [
+  kube.PersistentVolume(pv.name) {
+    metadata+: {
+      namespace: pv.namespace,
+    },
+    spec+: {
+      capacity: {
+        storage: pv.capacity,
+      },
+      accessModes: com.getValueOrDefault(pv, 'accessModes', [ 'ReadWriteMany' ],),
+      persistenVolumeReclaimPolicy: com.getValueOrDefault(pv, 'reclaimPolicy', null),
+      mountOptions: com.getValueOrDefault(pv, 'mountOptions', [ 'dir_mode=0777', 'file_mode=0777', 'vers=3.0' ]),
+      csi: {
+        driver: 'smb.csi.k8s.io',
+        readOnly: com.getValueOrDefault(pv, 'readOnly', null),
+        volumeAttributes: {
+          source: '//%s/%s' % [ pv.share_host, pv.share_name ],
+        },
+        volumeHandle: 'pv-%s-%s' % [ pv.namespace, pv.name ],
+        nodeStageSecretRef: {
+          name: '%s-credentials' % [ pv.name ],
+          namespace: pv.namespace,
+        },
+      },
+      storageClassName: 'smb-%s' % [ pv.namespace ],
+    },
+  }
+  for pv in params.pvs
+];
+
+local pvcs = [
+  kube.PersistentVolumeClaim(pv.name) {
+    metadata+: {
+      namespace: pv.namespace,
+    },
+    spec+: {
+      accessModes: com.getValueOrDefault(pv, 'accessModes', [ 'ReadWriteMany' ],),
+      resources: {
+        requests: {
+          storage: pv.capacity,
+        },
+      },
+      storageClassName: 'smb-%s' % [ pv.namespace ],
+      volumeMode: 'Filesystem',
+      volumeName: pv.name,
```

Comment:
AFAIK the `volumeName` is a readonly parameter that is set once the PVC is bound to a PV. In order to bind a PVC to a specifc PV, we should use [label selector](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#selector) instead. Otherwise, it could be that the PVC is bound to the wrong PV...
I suggest to try it out first or can you confirm that it works for you?

Reply:
The [spec](https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#persistentvolumeclaimspec-v1-core) doesn't say anything about `volumeName` being read-only. IIRC this should work correctly for the specific use-case that each SMB volume is preallocated to a namespace at creation time.

Reply:
I have already tried it and it works as expected.